### PR TITLE
Setup dependabot for updating pyright, pytype, flake8-pyi and GitHub Actions

### LINKED ISSUE #11484: Set up automation for updating pytype and pyright versions
Both pytype and pyright release pretty frequently and it's useful for us to make sure we test against the latest version. It would be nice to have a GitHub action that automatically updates them for us so we don't have to remember to make a PR. Example PRs: #11483, #11190.

I suppose other dependencies could also be auto-updated but the need isn't as great.

### PR DESCRIPTION
**Pull request by AlexWaygood**
Dependabot will do a daily check to see if there is a new version of pyright, pytype or flake8-pyi, and will create PRs updating them if so. It will also do a monthly check to see if there are any new versions for any of our GitHub actions, and will create a PR updating all of them together if so.

Dependabot will not check to see if any of our other dependencies are out of date. Ideally (in my opinion) it would do a monthly or quarterly check for those. Unfortunately, it doesn't look like it's possible to have two entries with different schedules for a single `package-ecosystem`.

In order to use dependabot with our pyright dependency, we switch to storing the pyright dependency in a minimal `package.json` file rather than as a field in our `pyproject.toml` file. This is the solution suggested by @jakebailey in https://github.com/python/typeshed/pull/11491#issuecomment-1972067313.

Fixes #11484
Closes #11491
Closes #11565

### CODE REVIEW

**Comment by AlexWaygood** [comment]
Ugh, seems like this breaks the stub-uploader...

https://github.com/python/typeshed/actions/runs/8223349890/job/22485946355?pr=11564

File: package.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    "name": "typeshed",
+    "version": "0.0.0",
```
Comment (AlexWaygood) [inline package.json:3]: These fields are required in order for this to be considered a valid `package.json` file: https://docs.npmjs.com/creating-a-package-json-file#packagejson-fields

**Comment by AlexWaygood** [comment]
> Ugh, seems like this breaks the stub-uploader, which depends on the `tool.typeshed.pyright_version` field of our `pyproject.toml`...
> 
> [python/typeshed/actions/runs/8223349890/job/22485946355?pr=11564](https://github.com/python/typeshed/actions/runs/8223349890/job/22485946355?pr=11564)

I fixed this failure by keeping the redundant `tool.typeshed.pyright_version` field in `pyproject.toml` for now, even though none of our tooling will actually use that field if this PR is merged. Once we've merged this PR, we can update the stub-uploader to read from the `package.json` file instead, and then we can remove the redundant field from our `pyproject.toml` file here.

**Comment by AlexWaygood** [comment]
> I'd prefer to keep all meta data in one file, instead of spreading it over multiple files.

That's definitely my preference too! But I'd also prefer to use widely used, well-tested tools for updating dependencies where possible. In this case, I _personally_ weigh that more heavily than my preference for keeping metadata consolidated.

But I agree this isn't ideal.

File: package.json (hunk 1 above)
Comment (Avasam) [reply package.json:3]: Just FYI if you ever want to add a comment to a package.json file, just add any top level field that starts with `//`. ie:
`"//": "my comment"`, `"// some-packge": "other comment about a package, with a unique comment key"` (the official recommendation is to make `"//"` as array, but that lacks comment positioning, so I've been doing it the way I described for years)
https://groups.google.com/g/nodejs/c/NmL7jdeuw0M/m/yTqI05DRQrIJ

File: .github/dependabot.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
+version: 2
```
Comment (Avasam) [inline .github/dependabot.yml:1]: Just so you know, dependabot will automatically add the [dependencies](https://github.com/python/typeshed/pulls?q=label%3Adependencies)  label with default configs (which I believe is something we'd want, but has only been used for a single PR on typeshed before)
https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#labels

**Review by Avasam — APPROVED** [review]
Looks good to me. I double checked, Dependabot seems to bee looking for any file that contains the string `requirement` in the name or any `.txt` file that passes their requirements parsing test, so it should pickup ours.
https://github.com/dependabot/dependabot-core/blob/main/python/lib/dependabot/python/file_fetcher.rb

As a bonus, we could get security advisory warnings about dependencies. Although typeshed stays pretty up to date so we're probably gonna outpace CVE publications.
https://docs.github.com/en/code-security/dependabot/dependabot-security-updates/configuring-dependabot-security-updates

**Comment by Avasam** [comment]
> I'm not super happy with this solution. I'd prefer to keep all meta data in one file, instead of spreading it over multiple files. I also don't think we should add a "fake" package.json file just to make dependabot happy. While it's more work, I'd prefer to use a custom solution for pyright or just keep updating it manually.

Thoughts on using https://pypi.org/project/pyright/ ?

**Review by Avasam — COMMENTED** [review]
Rescinding my approval because I just noticed this doesn't remove pyright version from pyproject.toml, which I think we could achieve.

**Comment by AlexWaygood** [comment]
I'm currently playing around with an alternative solution that uses renovate.

**Comment by jakebailey** [comment]
Yeah, if you're willing to use renovate, the regex manager can bump anything anywhere: https://docs.renovatebot.com/modules/manager/regex/

```toml
[tool.typeshed]
# renovate: datasource=npm depName=pyright
pyright_version = "1.1.350"
```

renovate ships with some presets, but you can make your own based on them, e.g. the GitHub Actions scanner that looks for variables named `*_VERSION`: https://docs.renovatebot.com/presets-regexManagers/#regexmanagersgithubactionsversions

**Comment by AlexWaygood** [comment]
See #11565 for what a solution using renovate might look like

**Comment by srittau** [comment]
> Thoughts on using https://pypi.org/project/pyright/ ? 

This sounds like the best solution to me. Are there potential problems with it? Otherwise, by just adding it to requirements, we'd also provide a neat way for testing locally.

If this would be problematic, using renovate also sounds promising.

**Pull request closed without merging**